### PR TITLE
Argument parsing for commands

### PR DESCRIPTION
**Pull request by MierenManz**
## About
Implements the feature request of #134

## Status
Not done!
- [x] These changes have been tested against Discord API or contain no API change.
- [x] This PR introduces some Breaking changes.

- [x] Add parser
- [x] Intergrate parser into message processing
- [ ] ~~Add dynamic return type~~

### CODE REVIEW

File: src/utils/command.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+export interface Args {
+  name: string
+  match: CommandArgumentMatchTypes
+  type?: unknown // Still needs to be implemented
+  defaultValue?: unknown // Still needs to be implemented
+  flag?: string
+}
+
+const mentionRegex = /([0-9]{18})/g
```
Comment (DjDeveloperr) [inline src/utils/command.ts:11]: Check `src/structures/messageMentions.ts` for regexes

**Comment by DjDeveloperr** [comment]
Suggestion: What about adding unit tests for args parsing? It would be useful for things like edge cases.

**Comment by MierenManz** [comment]
> Suggestion: What about adding unit tests for args parsing? It would be useful for things like edge cases.

I wouldn't really know how that test would look like?
Maybe add all 4 matches and supply some random string of input and assertEqual against what it should be?

File: src/utils/command.ts (hunk 1 above)
Comment (MierenManz) [reply src/utils/command.ts:11]: I'll definitely do that. right now I just wanted it to return the userID

**Comment by MierenManz** [comment]
> Suggestion: What about adding unit tests for args parsing? It would be useful for things like edge cases.

This is now added and I feel like the pr is ready for the final review now

**Review by Helloyunho — APPROVED** [review]
lgtm!

**Comment by MierenManz** [comment]
these were my last changes, imo ready to be landed 🚀

**Review by DjDeveloperr — APPROVED** [review]
LGTM! 🎉  I think `rest` would be more useful if it was string instead of string[]

**Comment by MierenManz** [comment]
> LGTM! 🎉 I think `rest` would be more useful if it was string instead of string[]

I can fix that for sure. then it's good to land I feel like

**Comment by DjDeveloperr** [comment]
> then it's good to land I feel like

Yes.
